cpp/include/raft/sparse/matrix/preprocessing.cuh

```diff
@@ -0,0 +1,490 @@
+ * @tparam IndexType
+ *   Type of the indices associated with the values.
+ *
+ * @param[in] vocab
+ *   Value that represents the number of features that exist for the matrices encoded.
+ */
+template <typename ValueType, typename IndexType>
+SparseEncoder<ValueType, IndexType>::SparseEncoder(int vocab) : vocabSize(vocab)
+{
+  cudaMallocManaged(&featIdCount, vocab * sizeof(int));
+  fullIdLen = 0;
+  numRows   = 0;
+  for (int i = 0; i < vocabSize; i++) {
+    featIdCount[i] = 0;
+  }
+}
+/**
+ * This constructor creates the `SparseEncoder` class with a vocabSize equal to the
+ * int vocab parameter supplied.
+ *
+ * @tparam ValueType
+ *   Type of the values in the sparse matrix.
+ * @tparam IndexType
+ *   Type of the indices associated with the values.
+ *
+ * @param[in] featIdValues
+ *   A map that consists of all the indices and values, to populate the featIdCount array.
+ * @param[in] num_rows
+ *   Value that represents the number of rows observed during fit cycle.
+ * @param[in] full_id_len
+ *   Value that represents the number overall number of features observed during the fit
+ * cycle.
+ * @param[in] vocab_size
+ *   Value that represents the number of features that exist for the matrices encoded.
+ * */
+template <typename ValueType, typename IndexType>
+SparseEncoder<ValueType, IndexType>::SparseEncoder(
+  raft::resources& handle, int* featIdValues, int vocab_size, int num_rows, int full_id_len)
+  : vocabSize(vocab_size), numRows(num_rows), fullIdLen(full_id_len)
+{
+  cudaStream_t stream = raft::resource::get_cuda_stream(handle);
+  cudaMallocManaged(&featIdCount, vocabSize * sizeof(int));
+  cudaMemset(featIdCount, 0, vocabSize * sizeof(int));
+  // raft::copy(featIdCount, featIdValues, vocabSize, stream);
+  for (int i = 0; i < vocabSize; i++) {
+    featIdCount[i] = featIdValues[i];
+  }
+}
+
+/**
+ * This destructor deallocates/frees the reserved memory of the class.
+ *
+ * @tparam ValueType
+ *   Type of the values in the sparse matrix.
+ * @tparam IndexType
+ *   Type of the indices associated with the values.
+ * */
+template <typename ValueType, typename IndexType>
+SparseEncoder<ValueType, IndexType>::~SparseEncoder()
+{
+  cudaFree(featIdCount);
+}
+
+template <typename ValueType, typename IndexType>
+void SparseEncoder<ValueType, IndexType>::save(raft::resources& handle, std::string save_path)
+{
+  cudaStream_t stream = raft::resource::get_cuda_stream(handle);
+  auto featIdCount_md = raft::make_device_vector<IndexType, int64_t>(handle, vocabSize);
+  raft::copy(featIdCount_md.data_handle(), featIdCount, vocabSize, stream);
+  std::ofstream saveFile(save_path);
+  if (saveFile.is_open()) {
+    std::ostringstream oss;
+    saveFile << vocabSize << " ";
+    saveFile << numRows << " ";
+    saveFile << fullIdLen << " ";
+    // serialize_mdspan<IndexType>(handle, oss, featIdCount_md.view());
+    for (int i = 0; i < vocabSize; i++) {
+      saveFile << featIdCount[i] << " ";
```

Comment:
could not use the serialize_mdspan, it does not work well when reading from a file that has other information on it.

Reply:
would probably be better to use a map here to save only indexes/values that are not zero. This would save time and space. Thoughts?

Reply:
fixed this, thanks for the tip